jax/interpreters/pxla.py

```diff
@@ -61,144 +61,138 @@
-      (i.e. sum(is_axis_materialized) == len(device_buffer.shape())). Any
-      unmaterialized axes must be sharded into size-1 chunks
-      (i.e. array_shape[i] == shards_per_axis[i]).
-    replication_factors: list of tuples of (factor, index) describing how many
-      times the array is replicated and before which logical axis index each
-      virtual replication axis is inserted.
+  `sharding` specifies how the array is supposed to get partitioned into chunks.
+  Its length should match the rank of the array. See the docstring of
+  `AvalDimSharding` for the supported partitioning schemes.
+
+  `mesh_mapping` describes an assignments of the array chunks created by `sharding`
+  to a logical device mesh. The length of the tuple is equal to the rank of the mesh.
+  Each mesh dimension can either get partitions of data varying along one of the
+  sharded dimensions, or the data can be replicated. See the docstring of
+  `MeshDimAssignment` for more information.
   """
+  sharding: Tuple[AvalDimSharding, ...]
+  mesh_mapping: Tuple[MeshDimAssignment, ...]
 
   def __init__(self,
-               shards_per_axis: Tuple[int, ...],
-               is_axis_materialized: Tuple[bool, ...],
-               replication_factors: List[Tuple[int, int]]):
-    assert len(shards_per_axis) == len(is_axis_materialized)
-    self.shards_per_axis = shards_per_axis
-    self.is_axis_materialized = is_axis_materialized
-    self.replication_factors = replication_factors
-
-  def __eq__(self, other):
-    return (self.shards_per_axis == other.shards_per_axis and
-            self.is_axis_materialized == other.is_axis_materialized and
-            self.replication_factors == other.replication_factors)
-
-  def __repr__(self):
-    return ("ShardingSpec(shards_per_axis=%s, is_axis_materialized=%s, "
-            "replication_factors=%s)" %
-            (self.shards_per_axis, self.is_axis_materialized,
-             self.replication_factors))
-
+               sharding: Iterable[AvalDimSharding],
+               mesh_mapping: Iterable[MeshDimAssignment]):
+    self.sharding = tuple(sharding)
+    self.mesh_mapping = tuple(mesh_mapping)
+
+  def indices(self, shape: Tuple[int, ...]) -> np.ndarray:
+    """Returns NumPy-style indices corresponding to a sharding spec.
+
+    Args:
+      shape: The shape of the logical array being sharded.
+
+    Returns:
+      An ndarray with a NumPy-style index for each element of the logical mesh.
+      Each element is an int, a slice object with step=1, or a tuple thereof, to
+      be treated as an index into the full logical array.
+    """
+    assert len(shape) == len(self.sharding)
+
+    axis_indices: List[Sequence[Index]] = []
+    chunk_indices_shape = []
+    for dim, sharding in enumerate(self.sharding):
+      axis_size = shape[dim]
+      if sharding is None:
+        axis_indices.append([slice(None)])
+        # NOTE: We don't append unsharded dimensions here, because they do not appear
```

Comment:
Yeah, `chunk_indices_shape`. I'll make it clearer in the comment

Reply:
Yeah, definitely. I didn't resolve all the comments just yet. Wanted to fold it into the final squash before the merge, because it seems to be a simple issue.